### DISCUSSION TITLE
Highest probability of topic not equal to topic selected by model?

### DISCUSSION THREAD

**Discussion by jermainkaminski** [MaartenGr/BERTopic#916] [Q&A] [answered]
I followed your [example](https://twitter.com/MaartenGr/status/1608127421919666176) to reduce outliers.

...

topics, probs = model.fit_transform(df.text)

% Reduce outliers

new_topics = model.reduce_outliers(df.text, topics, strategy="c-tf-idf")

%Update topic frequency

model.topics_ = new_topics
documents = pd.DataFrame({"Document": df.text, "Topic": new_topics})
model._update_topic_size(documents)

%Create a topics probabilities DataFrame

names = model.get_topic_info()['Topic']
prob_df = pd.DataFrame(probs)
prob_df = prob_df.rename(columns=names)
prob_df

%Get a dominant topic for each document

prob_df['dominant_topic'] = prob_df.columns[np.argmax(prob_df.values, axis=1)]
prob_df['topic'] = model.topics_
prob_df

Here, I realized that the _np.argmax_ value of the row is not always equal to the __topic_ value assigned by the model. In other words, the column with the highest probability is not always the assigned topic.

<img width="509" alt="Screenshot 2023-01-09 at 18 02 38" src="https://user-images.githubusercontent.com/6346677/211365064-bcb352b5-2a60-411b-a974-8845cc6e41e9.png">

Is that because the _probs_ were not updated with _reduce_outliers_, or is there an alternative explanation?
In case, it would be great to know how to update probabilities once outliers are reduced (and to fix the issue).

Thank you in advance!

**Comment by MaartenGr** [comment, marked as answer]
The probabilities that are generated within BERTopic are a result of the soft clustering capabilities of HDBSCAN. More specifically, these probabilities are generated **after** finding the clusters and are exactly an approximation of the probabilities, not the process by which the clusters are created. In other words, the probabilities that are calculated by HDBSCAN are an approximation of its fitting process and will not exactly match the actual clusters that are created. 

Lastly, the output of `.reduce_outliers` is done with a different process than how the probabilities are calculated, so updating the probabilities with the output of `.reduce_outliers` is not possible since they are created from different methods.

**Reply by jermainkaminski** [reply]
Thank you, @MaartenGr! This is very helpful.

Just asking, as the reproducibility of research papers might depend on this: Is there any workaround you'd recommend? Or: Would you rather recommend to work with the `np.argmax` of a row, or the respective `_topic` assigned?

**Reply by MaartenGr** [reply]
Working with the topics themselves is, in the case of using HDBSCAN, often preferred. The reason for this is that the probabilities that you extract using HDBSCAN are an approximation, and as such are less accurate than the actual clustering itself since it tries to imitate the clustering in a sense. You can find more about that [here](https://hdbscan.readthedocs.io/en/latest/soft_clustering_explanation.html). However, the soft probabilities are especially helpful if you have a bunch of outliers that you want to assign to topics or if you want to know more about the potential topic distributions.
